File: workstation/src/main/java/io/xj/engine/fabricator/SegmentRetrospective.java
// Copyright (c) XJ Music Inc. (https://xjmusic.com) All Rights Reserved.
package io.xj.engine.fabricator;

import io.xj.model.enums.InstrumentMode;
import io.xj.model.enums.InstrumentType;
import io.xj.model.enums.ProgramType;
import io.xj.engine.FabricationException;
import io.xj.model.pojos.Segment;
import io.xj.model.pojos.SegmentChoice;
import io.xj.model.pojos.SegmentChoiceArrangement;
import io.xj.model.pojos.SegmentChoiceArrangementPick;
import io.xj.model.pojos.SegmentChord;
import io.xj.model.pojos.SegmentMeta;

import java.util.Collection;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

/**
 Digest segments of the previous main program
 <p>
 NextMain/NextMacro-type: Retrospective of the previous main choice, primary choices only
 REF https://github.com/xjmusic/workstation/issues/242
 <p>
 Continue-type: Retrospective of all segments in this main program
 REF https://github.com/xjmusic/workstation/issues/242
 */
public interface SegmentRetrospective {

  /**
   Get the arrangement for the given pick

   @param pick for which to get arrangement
   @return arrangement
   @throws FabricationException on failure to retrieve
   */
  SegmentChoiceArrangement getArrangement(SegmentChoiceArrangementPick pick) throws FabricationException;

  /**
   @return all choices
   */
  Collection<SegmentChoice> getChoices();

  /**
   Get the choice for the given arrangement

   @param arrangement for which to get choice
   @return choice
   @throws FabricationException on failure to retrieve
   */
  SegmentChoice getChoice(SegmentChoiceArrangement arrangement) throws FabricationException;

  /**
   Get the instrument type for a given pick

   @param pick for which to get instrument type
   @return instrument type of pick
   */
  InstrumentType getInstrumentType(SegmentChoiceArrangementPick pick) throws FabricationException;

  /**
   Get the meta from the previous segment with the given key
   <p>
   Segment has metadata for XJ to persist "notes in the margin" of the composition for itself to read https://github.com/xjmusic/workstation/issues/222

   @param key to search for meta
   @return meta if found
   */
  Optional<SegmentMeta> getPreviousMeta(String key);

  /**
   Get the previous segment choices for the given instrument
   (although there should only be one previous segment choice for each instrument)

   @param instrumentId for which to get choice
   @return previous segment choice
   */
  Collection<SegmentChoice> getPreviousChoicesForInstrument(UUID instrumentId);

  /**
   Get the previous arrangements for the given instrument id

   @param instrumentId for which to get arrangements
   @return segment choice arrangements
   */
  Collection<SegmentChoiceArrangement> getPreviousArrangementsForInstrument(UUID instrumentId);

  /**
   Get the picks of any previous segments which selected the same main sequence
   <p>
   Artist writing detail program expects 'X' note value to result in random part creation from available Voicings https://github.com/xjmusic/workstation/issues/251

   @return map of all previous segment meme constellations (as keys) to a collection of choices made
   */
  Collection<SegmentChoiceArrangementPick> getPicks();

  /**
   Get the choice of a given type

   @param type of choice to get
   @return choice of given type
   */
  Optional<SegmentChoice> getPreviousChoiceOfType(Segment segment, ProgramType type);

  /**
   Get the previous-segment choice of a given type

   @param type of choice to get
   @return choice of given type
   */
  Optional<SegmentChoice> getPreviousChoiceOfType(ProgramType type);

  /**
   Get the previous-segment choices of a given instrument mode

   @param instrumentMode for which to get previous-segment choices
   @return choices
   */
  List<SegmentChoice> getPreviousChoicesOfMode(InstrumentMode instrumentMode);

  /**
   Get the previous-segment choices of a given instrument type and mode

   @param instrumentType  for which to get previous-segment choices
   @param instrumentModes for which to get previous-segment choices
   @return choices
   */
  List<SegmentChoice> getPreviousChoicesOfTypeMode(InstrumentType instrumentType, InstrumentMode instrumentModes);

  /**
   Get the previous-segment choices of a given instrument type

   @param instrumentType for which to get previous-segment choices
   @return choices
   */
  Optional<SegmentChoice> getPreviousChoiceOfType(InstrumentType instrumentType);

  /**
   Get the previous picks for the given instrument id

   @param instrumentId for which to get picks
   @return segment choice picks
   */
  Collection<SegmentChoiceArrangementPick> getPreviousPicksForInstrument(UUID instrumentId);

  /**
   Get the segment immediately previous to the current segment

   @return previous segment
   */
  Optional<Segment> getPreviousSegment();

  /**
   @return all cached segments
   <p>
   Always starts with the segment at current offset minus one,
   includes that segment and all others with the same main program, which covers both
   whether this retrospective is looking back on the current main program (Continue segment),
   or the previous one (NextMain/NextMacro segments)
   */
  Collection<Segment> getSegments();

  /**
   Get all segment chords for the given segment id

   @param segmentId for which to get chords
   @return chords
   */
  List<SegmentChord> getSegmentChords(int segmentId);

}

File: engine/src/fabricator/SegmentRetrospective.cpp
// Copyright (c) XJ Music Inc. (https://xjmusic.com) All Rights Reserved.

#include <utility>

#include "xjmusic/fabricator/FabricationFatalException.h"
#include "xjmusic/fabricator/SegmentRetrospective.h"
#include "xjmusic/fabricator/FabricationException.h"

using namespace XJ;

SegmentRetrospective::SegmentRetrospective(SegmentEntityStore *entityStore, const int segmentId, const bool autoload) {
  this->entityStore = entityStore;
  this->segmentId = segmentId;

  // NOTE: the segment retrospective is empty for segments of type Initial, NextMain, and NextMacro--
  // Only Continue-type segments have a retrospective
  // begin by getting the previous segment
  // only can build retrospective if there is at least one previous segment
  // the previous segment is the first one cached here. we may cache even further back segments below if found
  if (!autoload || segmentId <= 0) {
    retroSegments = std::vector<const Segment *>();
    previousSegmentIds = std::set<int>();
    previousSegment = std::nullopt;
    return;
  }

  // begin by getting the previous segment
  // the previous segment is the first one cached here. we may cache even further back segments below if found
  previousSegment = entityStore->readSegment(segmentId - 1);
  if (!previousSegment.has_value()) {
    throw FabricationFatalException("Retrospective sees no previous segment!");
  }
  segmentChords.emplace(previousSegment.value()->id, entityStore->readOrderedSegmentChords(previousSegment.value()->id));

  // previous segment must have a main choice to continue past here.
  const std::optional<const SegmentChoice *> previousSegmentMainChoice =
      entityStore->readChoice(previousSegment.value()->id, Program::Type::Main);

  if (!previousSegmentMainChoice.has_value() || previousSegmentMainChoice.value()->programType != Program::Type::Main) {
    throw FabricationFatalException("Retrospective sees no main choice!");
  }

  retroSegments = entityStore->readAllSegments();

  for (const auto &s: retroSegments) {
    auto c = entityStore->readChoice(s->id, Program::Type::Main);
    if (c.has_value() && previousSegmentMainChoice.value()->programId == c.value()->programId) {
      previousSegmentIds.emplace(s->id);
      segmentChords.emplace(s->id, entityStore->readOrderedSegmentChords(s->id));
    }
  }
}

std::optional<const SegmentChoice *>
SegmentRetrospective::getPreviousChoiceOfType(const Segment *segment, const Program::Type programType) const {
  const auto c = entityStore->readChoice(segment->id, programType);
  return c.has_value() && programType == c.value()->programType ? c : std::nullopt;
}

std::set<const SegmentChoiceArrangementPick *> SegmentRetrospective::getPicks() const {
  // return new ArrayList<>(retroStore.getAll(SegmentChoiceArrangementPick.class));
  return entityStore->readAllSegmentChoiceArrangementPicks(previousSegmentIds);
}

std::optional<const Segment *> SegmentRetrospective::getPreviousSegment() const {
  return previousSegment;
}

std::optional<const SegmentChoice *> SegmentRetrospective::getPreviousChoiceOfType(const Program::Type programType) const {
  if (!previousSegment.has_value()) return std::nullopt;
  return getPreviousChoiceOfType(previousSegment.value(), programType);
}

std::set<const SegmentChoice *> SegmentRetrospective::getPreviousChoicesOfMode(const Instrument::Mode instrumentMode) const {
  std::set<const SegmentChoice *> result;
  if (!previousSegment.has_value()) return result;
  const auto choices = entityStore->readAllSegmentChoices(previousSegment.value()->id);
  for (const auto &choice: choices) {
    if (choice->instrumentMode == instrumentMode) {
      result.emplace(choice);
    }
  }
  return result;
}

std::set<const SegmentChoice *>
SegmentRetrospective::getPreviousChoicesOfTypeMode(const Instrument::Type instrumentType, const Instrument::Mode instrumentMode) const {
  std::set<const SegmentChoice *> result;
  if (!previousSegment.has_value()) return result;
  const auto choices = entityStore->readAllSegmentChoices(previousSegment.value()->id);
  for (const auto &choice: choices) {
    if (choice->instrumentMode == instrumentMode && choice->instrumentType == instrumentType) {
      result.emplace(choice);
    }
  }
  return result;
}

std::optional<const SegmentChoice *> SegmentRetrospective::getPreviousChoiceOfType(const Instrument::Type instrumentType) const {
  if (!previousSegment.has_value()) return std::nullopt;
  const auto choices = entityStore->readAllSegmentChoices(previousSegment.value()->id);
  for (const auto &choice: choices) {
    if (choice->instrumentType == instrumentType) {
      return choice;
    }
  }
  return std::nullopt;
}

std::vector<const Segment *> SegmentRetrospective::getSegments() const {
  return retroSegments;
}

std::set<const SegmentChoice *> SegmentRetrospective::getChoices() const {
  return entityStore->readAllSegmentChoices(previousSegmentIds);
}

std::set<const SegmentChoice *> SegmentRetrospective::getPreviousChoicesForInstrument(const UUID &instrumentId) const {
  std::set<const SegmentChoice *> result;
  if (!previousSegment.has_value()) return result;
  const auto choices = entityStore->readAllSegmentChoices(previousSegment.value()->id);
  for (const auto &choice: choices) {
    if (choice->instrumentId == instrumentId) {
      result.emplace(choice);
    }
  }
  return result;
}

std::set<const SegmentChoiceArrangement *> SegmentRetrospective::getPreviousArrangementsForInstrument(const UUID &instrumentId) const {
  std::set<const SegmentChoiceArrangement *> result;
  const auto choices = getPreviousChoicesForInstrument(instrumentId);
  const auto arrangements = entityStore->readAllSegmentChoiceArrangements(previousSegmentIds);
  for (const auto &choice: choices) {
    for (const auto &arrangement: arrangements) {
      if (arrangement->segmentChoiceId == choice->id) {
        result.emplace(arrangement);
      }
    }
  }
  return result;
}

std::set<const SegmentChoiceArrangementPick *> SegmentRetrospective::getPreviousPicksForInstrument(const UUID &instrumentId) const  {
  std::set<const SegmentChoiceArrangementPick *> result;
  if (!previousSegment.has_value()) return result;
  const auto arrangements = getPreviousArrangementsForInstrument(std::move(instrumentId));
  const auto picks = entityStore->readAllSegmentChoiceArrangementPicks(
      previousSegmentIds);
  for (const auto &arrangement: arrangements) {
    for (const auto &pick: picks) {
      if (pick->segmentChoiceArrangementId == arrangement->id) {
        result.emplace(pick);
      }
    }
  }
  return result;
}

Instrument::Type SegmentRetrospective::getInstrumentType(const SegmentChoiceArrangementPick *pick) const  {
  return getChoice(getArrangement(pick))->instrumentType;
}

std::optional<const SegmentMeta *> SegmentRetrospective::getPreviousMeta(const std::string &key) const  {
  const auto metas = entityStore->readAllSegmentMetas(previousSegmentIds);
  for (const auto &meta: metas) {
    if (meta->key == key) {
      return meta;
    }
  }
  return std::nullopt;
}

const SegmentChoiceArrangement * SegmentRetrospective::getArrangement(const SegmentChoiceArrangementPick *pick) const  {
  const auto arrangements = entityStore->readAllSegmentChoiceArrangements(pick->segmentId);
  for (const auto &arrangement: arrangements) {
    if (arrangement->id == pick->segmentChoiceArrangementId) {
      return arrangement;
    }
  }
  throw FabricationException("Failed to get arrangement for SegmentChoiceArrangementPick[" + pick->id + "]");
}

const SegmentChoice *SegmentRetrospective::getChoice(const SegmentChoiceArrangement *arrangement) const  {
  const auto choices = entityStore->readAllSegmentChoices(arrangement->segmentId);
  for (const auto &choice: choices) {
    if (choice->id == arrangement->segmentChoiceId) {
      return choice;
    }
  }
  throw FabricationException("Failed to get arrangement for SegmentChoiceArrangement[" + arrangement->id + "]");
}

std::vector<const SegmentChord *> SegmentRetrospective::getSegmentChords(const int segmentId) const  {
  if (segmentChords.find(segmentId) == segmentChords.end()) {
    return std::vector<const SegmentChord *>();
  }
  return segmentChords.at(segmentId);
}

